File: internal/app/repo/postgres/transaction_repo.go
package postgres

import (
	"context"
	"database/sql"
	"log/slog"
	"mahir-trade-be/internal/app/models"
	"mahir-trade-be/internal/app/repo/postgres/queries"

	"go.uber.org/dig"
)

type (
	TransactionRepo interface {
		CreateTransaction(ctx context.Context, req models.Transaction) (id int, err error)
	}

	TransactionRepoImpl struct {
		dig.In

		*sql.DB
	}
)

func NewTransactionRepo(impl TransactionRepoImpl) TransactionRepo {
	return &impl
}

func (t *TransactionRepoImpl) CreateTransaction(ctx context.Context, req models.Transaction) (id int, err error) {
	rows, err := t.QueryContext(ctx, queries.QueryCreateTransaction, req.OrderID, req.Amount, req.SettlementDate, req.WebhookID, req.CreatedBy)
	if err != nil {
		slog.ErrorContext(ctx, "[transactionRepoImpl][CreateTransaction] error while QueryContext", "%v", err.Error())
		return id, err
	}

	defer rows.Close()

	for rows.Next() {
		err = rows.Scan(&id)
		if err != nil {
			slog.ErrorContext(ctx, "[transactionRepoImpl][CreateTransaction] error while Scan", "%v", err.Error())
			return id, err
		}
	}

	return id, nil
}

File: src/repo/transaction_repo.ts
import { Pool } from "pg";
import { Transaction } from "../models/transaction.model";

import db from "../db/db.config";
import { TransactionQueries } from "./queries/transaction_queries";

export class TransactionRepository {
  private pg: Pool;

  constructor() {
    this.pg = db;
  }

  // --- CREATE TRANSACTION ---
  async createTransaction(req: Transaction): Promise<number> {
    try {
      const result = await this.pg.query(TransactionQueries.CreateTransaction, [
        req.order_id,
        req.amount,
        req.settlement_date,
        req.webhook_id,
        req.created_by,
      ]);

      if (result.rowCount === 0) {
        throw new Error("failed to create transaction");
      }

      return result.rows[0].id as number;
    } catch (err: any) {
      console.error(
        `[TransactionRepository][createTransaction] error: ${err.message}`
      );
      throw err;
    }
  }
}
